plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs

```diff
@@ -198,45 +208,53 @@ unRenamedE dflags expr = do
 -- TODO: workaround when HieAst unavailable (e.g. when the module itself errors)
 -- TODO: Declaration Splices won't appear in HieAst; perhaps we must just use Parsed/Renamed ASTs?
 codeAction :: CodeActionProvider IdeState
-codeAction _ state plId docId range0 _ =
+codeAction _ state plId docId ran _ =
     fmap (maybe (Right $ List []) Right) $
         runMaybeT $ do
             fp <- MaybeT $ pure $ uriToNormalizedFilePath $ toNormalizedUri theUri
-            (getAsts . hieAst -> asts, posMap) <-
+            ParsedModule {..} <-
                 MaybeT . runAction "splice.codeAction.GitHieAst" state $
-                    useWithStale GetHieAst fp
-            ran' <-
-                MaybeT $
-                    pure $
-                        fromCurrentRange posMap range0
-            fmap (List . DL.toList) $
-                execWriterT $
-                    getAp $ ifoldMap (fmap Ap . go ran') asts
+                    use GetParsedModule fp
+            let spn =
+                    rangeToRealSrcSpan ran $
+                        fromString $
+                            fromNormalizedFilePath fp
+                mouterSplice = something' (detectSplice spn) pm_parsed_source
```

Comment:
Why not just check if `spn` is inside one of the spans of the unexpanded splices? (i.e. `spn` lies within one of the Locateds of the `LHsExpr GhcTc`)

Reply:
Because there can be nested splices in practice and hence we must choose the nnermost one for disambiguation:

```hs
a x = putStrLn $([|"foo" + $(varE $ mkName "x")|])
```

Reply:
You can pick the smallest one from the list of `LHsExpr GhcTc` that fits. I don't see the need for a syb traversal when we've already captured all the information we need.

Reply:
For the splices that passed kind-checking, using TypeCheck result makes sense.
But, if one wants to detect all the splice, we need to analyse through the AST.
I confirmed this by experimenting by feeding the following:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TemplateHaskell #-}
module TTypeKindError where
import Language.Haskell.TH ( numTyLit, litT )
import Data.Proxy ( Proxy )

main :: $(litT (numTyLit 42))
main = return ()
```

Using raw `ParsedModule`, `$(litT ...)` part will be detected. But when using `tmrToplevelSplices`, it won't be detected.

Reply:
Another concern to made me using `ParsedModule` here was that type-checking could take much longer than just parsing, so I think it might be faster to use in CodeAction detection phase than requesting `TypeCheck`.

Reply:
How will you expand it if it is not detected?

Also, you can probably save and return the splice information even if the module fails to typecheck - this will require some wrangling though to figure out where and how to return this information.

> so I think it might be faster to use in CodeAction detection phase than requesting TypeCheck

If you `useWithStale`, this will not be a concern in 99% of cases

Reply:
> How will you expand it if it is not detected?

In some cases, my first implementation should work. For example, the above example won't kind-check as a whole, but the splice `$(litT (numTyLit 42))` itself can be expanded if it is considered standalone; and users should expect them to be expanded when debugging.
We can apply functions from `RnSplice` to such examples.

> Also, you can probably save and return the splice information even if the module fails to typecheck - this will require some wrangling though to figure out where and how to return this information.

Yes, it might be helpful to store expanded results in `ExpandSpliceParams`, perhaps wrapped in Maybe.

> If you useWithStale, this will not be a concern in 99% of cases

Sounds good. How likely is it possible that staled module could differ from the current state and destroys the source entirely?
During developing this plugin, I've encountered the situation that Tactic Plugin destroys some lines.

Reply:
> Sounds good. How likely is it possible that staled module could differ from the current state and destroys the source entirely?

If you `useWithStale` to generate the code action suggestions and then do proper `use` to generate the actual edit, I don't think we will ever mangle the source file.

Reply:
I decided to keep using ParsedModule in splice detection, since, as mentioned in https://github.com/haskell/haskell-language-server/pull/759#issuecomment-753313091, now we can expand some splices faild due to type- or syntactic-error given that staled TcModuleResult is available.

> Also, you can probably save and return the splice information even if the module fails to typecheck - this will require some wrangling though to figure out where and how to return this information.

I will try this approach in another p.r., if we can find the way to achieve this!